File: Leetcode/PY/Leetcode911-在线选举.py
class TopVotedCandidate:
    def __init__(self, persons: List[int], times: List[int]):
        n = len(persons)
        self.win = [0] * n
        self.time = times
        ss = [0] * n
        maxc ,maxp = 0,-1
        for i in range(n):
            p = persons[i]
            ss[p] += 1
            if ss[p] >= maxc:
                maxc = ss[p]
                maxp = p
            self.win[i] = maxp

    def q(self, t: int) -> int:
        l,r = 0,len(self.win) - 1
        while l < r:
            mid = (l + r + 1) // 2
            if self.time[mid] <= t: l = mid
            else: r = mid - 1
        return self.win[l]

File: Leetcode/JS/Leetcode911-在线选举.js
/**
 * @param {number[]} persons
 * @param {number[]} times
 */
 var TopVotedCandidate = function(persons, times) {
    let n = persons.length;
    this.win = new Array(n).fill(0);
    this.t = times;
    let cnt = new Array(n).fill(0);
    let mxc = 0,mxp = -1;
    for(let i = 0;i < n;i++){
        let p = persons[i];
        ++cnt[p];
        if(cnt[p] >= mxc) {
            mxp = p;
            mxc = cnt[p];
        }
        this.win[i] = mxp;
    }
    // for(let i = 0;i < n;i++) console.log(this.win[i]);
};

/** 
 * @param {number} t
 * @return {number}
 */
TopVotedCandidate.prototype.q = function(t) {
    let l = 0,r = this.t.length - 1;
    while(l < r) {
        let mid = Math.floor((l + r + 1) / 2);
        if(this.t[mid] <= t) l = mid;
        else r = mid - 1;
    }
    return this.win[l];
};
